### ISSUE TITLE
feat: block-stripped pytree_eval_fn for factorize_axes vmap path

### ISSUE THREAD

**Issue by jc-macdonald** [ACCIDDA/op_system#168]
## Summary

When `factorize_axes: [<axis>]` is declared, `diffrax_engine` runs `jax.vmap(pytree_eval_fn, in_axes=<axis_pos>)` to evaluate the ODE per-block. But `pytree_eval_fn` is compiled against the **full** state shape (block axis still present), so its lowered IR contains literal axis indices that assume the block axis exists.

Under vmap, the function is traced on per-block slices (block axis removed). Any alias compiled to `np.transpose(buf, (2, 0, 1))` or `np.sum(buf, axis=k_loc)` then operates on an array of lower rank than expected and crashes with `axis 2 out of bounds for array of dimension 2` (or shape-mismatch).

This affects **every** spec that has aliases or operators referencing the block axis when `factorize_axes` is active, not just the COVID19_USA configs that triggered the discovery.

## Reproduction

Any `pytree_mode` spec with `factorize_axes: [loc]` and an alias like
```yaml
aliases:
  N_total[loc]: apply_along(E[age,loc] + I[age,loc], age=a, kernel=sum)
```
will fail at solver tracing time with an axis-index error from inside the compiled alias body.

## Root cause

`pytree_eval_fn` is **rank-rigid**. The lowering in `_ir_lower.py` and the CSE in `_vectorize.py` bake literal axis positions (`np.transpose(buf, (2, 0, 1))`, `np.sum(buf, axis=2)`, etc.) into the closure. When the engine vmaps over the block axis, vmap traces the function on slices and the literal indices reference axis positions that no longer exist.

`analyze_block_axes` already proves the block axis is structurally separable from the IR, so producing a block-stripped version of the IR (and lowering it) is well-defined.

## Decision

Add a second compiled function on `CompiledRhs` that operates on the **block-stripped** state shape. The engine swaps `pytree_eval_fn → block_pytree_eval_fn` when entering the per-block vmap path.

Rejected alternatives:

- **Rank-polymorphic lowering** (rewrite `_ir_lower.py` to never emit literal axis indices). Invasive; some `apply_along` chains genuinely require literal transposes; risks performance regression on the non-block path.
- **Spec-text preprocessing** (delete the block axis from subscripts in the raw spec before normalize). Brittle (alias bodies are free-form), doesn't handle parameters with a block axis, and doesn't compose with multiple block axes.
- **Fix in `diffrax_engine`** (current state). Couples engine to op_system IR internals; doesn't work for non-trivial aliases; we already have `_build_operator_terms_pytree_block` doing axis arithmetic that is exactly the symptom we want to remove.

## Proposal — Option 1: block-stripped compile pass

Add an additive field to `CompiledRhs`:
```python
@dataclass(frozen=True, slots=True)
class CompiledRhs:
    ...
    block_pytree_eval_fn: PytreeEvalFn | None = None     # NEW
    block_template_shapes: dict[str, tuple[int, ...]] | None = None  # NEW
    # block_axes: tuple[BlockAxisInfo, ...]  # already exists
```

When `block_axes` is non-empty (and there is exactly one for now — multi-axis is a follow-up), `compile_rhs` performs a second normalize → lower → vectorize pass against a **block-stripped** IR view:

1. `strip_block_axis(rhs_ir, axis_name)` in a new `_normalize_block.py`:
   - Walks the IR.
   - Deletes `axis_name` from every state template shape and alias shape.
   - Removes coord-pinned subscripts on `axis_name` (e.g. `[loc:l]` → bare).
   - Drops `Reduce` nodes whose axis is `axis_name` (already proven unused by `analyze_block_axes`).
   - Strips `axis_name` from time-varying / shaped parameter axes.
   - Refuses to strip if `analyze_block_axes` was not run or did not approve the axis.

2. `compile_rhs` (in `_vectorize.py`):
   - After the existing flat + pytree compile, if `block_axes` is non-empty, call `strip_block_axis` and run `lower_to_vector_ast` + the existing eval-fn builder on the stripped IR.
   - Attach as `block_pytree_eval_fn` and `block_template_shapes`.

3. `diffrax_engine.py`:
   - In `_solve_one_block`, call `compiled.block_pytree_eval_fn` instead of `compiled.pytree_eval_fn`.
   - Delete `_build_operator_terms_pytree_block` (its job moves into op_system).
   - In `_prepare_call`, when picking `block_axis`, also pin `pytree_stepper_fn = compiled.block_pytree_eval_fn` for the inner vmap'd path.

## Acceptance criteria

- [ ] All existing op_system tests pass.
- [ ] New tests in `test_block_axes.py`:
  - `test_block_stripped_eval_fn_matches_monolithic` — for a spec with an alias `N_total[loc] = sum_over(age, vax)`, `compiled.pytree_eval_fn(y_full)[base][:, :, i]` ≈ `vmap(compiled.block_pytree_eval_fn)(y_full)[base][i]` for every block coord `i`.
  - `test_block_stripped_aliases_have_correct_rank` — the stripped IR's aliases have no `Reduce` over the block axis and no subscripts mentioning it.
  - `test_block_stripped_tv_params_axes` — `beta[time, loc]` becomes `beta[time]` in the stripped param shapes.
- [ ] `_build_operator_terms_pytree_block` removed from `diffrax_engine.py` (in COVID19_USA repo; PR will land there separately).
- [ ] End-to-end: prior-predictive of `SMH_R19_op_system_hierarchical` with `factorize_axes: [loc]` runs to completion and matches the monolithic-path result within solver tolerance.

## Files

| File | Change |
|---|---|
| `src/op_system/_normalize_block.py` (new) | `strip_block_axis(rhs, axis_name) -> NormalizedRhs` |
| `src/op_system/_vectorize.py` | Second compile pass when `block_axes` non-empty; attach `block_pytree_eval_fn`, `block_template_shapes` |
| `src/op_system/compile.py` | Surface new fields on `CompiledRhs` |
| `src/op_system/_block_axes.py` | Optional: expose helper enumerating axis positions in the *stripped* template shapes (for engine `in_axes`) |
| `flepimop2-op_system/src/.../__init__.py` | Surface `block_template_shapes` on connector options |
| `tests/op_system/test_block_axes.py` | New tests above |
| `docs/development/block-axis-vmap-plan.md` | Append "Phase 2: block-stripped compile" section |

## Out of scope

- Multi-block-axis stripping (rejected here as YAGNI; one axis covers the spatial-factorization use case).
- Block-axis vmap nesting (handled by the engine, unchanged).
- Naming change of `factorize_axes` (separate discussion).

## Discovery

Hit while running `flepimop2 process configs/SMH_R19_op_system_hierarchical.yml -t prior_predictive` after Slice D (#162 follow-on) landed in COVID19_USA. Stack trace ends in `alias 'I_total' evaluation failed: ValueError('axis 2 is out of bounds for array of dimension 2')` from inside the vmapped `pytree_stepper_fn`. Workaround in the engine (`_build_operator_terms_pytree_block`) covers operators but not aliases.

Related: #156 (typed `OperatorDescriptor`), #158 (PyTree state interface), #162 (Slice A: `analyze_block_axes`).
